File: rust/ems-core/src/journal.rs
//! The JSONL event journal — the wire format the conformance gate compares byte-for-byte.

use std::collections::BTreeMap;
use std::fmt;
use std::fmt::Write as _;
use std::fs;
use std::io;
use std::path::Path;

const KEY_FIELDS: &str = "fields";
const KEY_SEQ: &str = "seq";
const KEY_TYPE: &str = "type";

/// One line of an event journal.
///
/// `fields` is a [`BTreeMap`] rather than a `HashMap` on purpose: its iteration
/// order reaches the output journal, and the conformance gate compares that
/// journal byte-for-byte across three languages. `clippy.toml` bans `HashMap`
/// workspace-wide for the same reason.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JournalEvent {
    /// Monotonically increasing sequence number, starting at 1.
    pub seq: u64,
    /// Event type name, e.g. `OrderNew`.
    pub event_type: String,
    /// Event payload; keys iterate in lexicographic order.
    pub fields: BTreeMap<String, String>,
}

impl JournalEvent {
    /// Creates an event with no payload.
    #[must_use]
    pub fn of(seq: u64, event_type: &str) -> Self {
        Self {
            seq,
            event_type: event_type.to_owned(),
            fields: BTreeMap::new(),
        }
    }

    /// Returns a copy of this event carrying a different sequence number.
    #[must_use]
    pub fn with_seq(&self, seq: u64) -> Self {
        Self {
            seq,
            ..self.clone()
        }
    }
}

/// A journal line could not be parsed.
///
/// Always carries the 1-based line number. The journal parser is one of the
/// three fuzz targets in the polyglot gate, so every malformed input must
/// produce this — never a panic, never a silently skipped line.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MalformedJournal {
    /// 1-based line number the failure was found on.
    pub line: usize,
    /// What was wrong.
    pub message: String,
}

impl fmt::Display for MalformedJournal {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "line {}: {}", self.line, self.message)
    }
}

impl std::error::Error for MalformedJournal {}

/// Anything that can go wrong reading or writing a journal.
#[derive(Debug)]
pub enum JournalError {
    /// The file could not be read or written.
    Io(io::Error),
    /// A line could not be parsed.
    Malformed(MalformedJournal),
}

impl fmt::Display for JournalError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(e) => write!(f, "{e}"),
            Self::Malformed(e) => write!(f, "{e}"),
        }
    }
}

impl std::error::Error for JournalError {}

impl From<io::Error> for JournalError {
    fn from(e: io::Error) -> Self {
        Self::Io(e)
    }
}

impl From<MalformedJournal> for JournalError {
    fn from(e: MalformedJournal) -> Self {
        Self::Malformed(e)
    }
}

// ── writing ──────────────────────────────────────────────────────────────────

/// Writes every event, one per line, replacing any existing file.
///
/// # Errors
/// Returns [`JournalError::Io`] if the file cannot be written.
pub fn write_journal(path: &Path, events: &[JournalEvent]) -> Result<(), JournalError> {
    let mut out = String::new();
    for event in events {
        encode_into(&mut out, event);
        out.push('\n');
    }
    fs::write(path, out.as_bytes())?;
    Ok(())
}

/// Encodes one event, without the trailing newline.
#[must_use]
pub fn encode(event: &JournalEvent) -> String {
    let mut out = String::new();
    encode_into(&mut out, event);
    out
}

fn encode_into(out: &mut String, event: &JournalEvent) {
    out.push_str("{\"");
    out.push_str(KEY_FIELDS);
    out.push_str("\":{");
    for (i, (key, value)) in event.fields.iter().enumerate() {
        if i > 0 {
            out.push(',');
        }
        append_string(out, key);
        out.push(':');
        append_string(out, value);
    }
    out.push_str("},\"");
    out.push_str(KEY_SEQ);
    out.push_str("\":");
    out.push_str(&event.seq.to_string());
    out.push_str(",\"");
    out.push_str(KEY_TYPE);
    out.push_str("\":");
    append_string(out, &event.event_type);
    out.push('}');
}

fn append_string(out: &mut String, value: &str) {
    out.push('"');
    for c in value.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{8}' => out.push_str("\\b"),
            '\u{c}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => {
                // No short escape exists; a raw control byte in the journal is a
                // byte-level divergence waiting to happen between three writers.
                // write! into the buffer rather than allocating a temporary String
                // per control character.
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            // Non-ASCII is written as UTF-8, not escaped. All three languages
            // agree on UTF-8; they would not agree on when to escape.
            c => out.push(c),
        }
    }
    out.push('"');
}

// ── reading ──────────────────────────────────────────────────────────────────

/// Reads every non-blank line.
///
/// # Errors
/// Returns [`JournalError::Io`] if the file cannot be read, or
/// [`JournalError::Malformed`] on the first line that does not parse.
pub fn read_journal(path: &Path) -> Result<Vec<JournalEvent>, JournalError> {
    let raw = fs::read_to_string(path)?;
    let mut events = Vec::new();
    for (index, line) in raw.lines().enumerate() {
        if line.trim().is_empty() {
            continue;
        }
        events.push(decode(line, index + 1)?);
    }
    Ok(events)
}

/// Decodes one line. `line_number` is 1-based and appears in any error message.
///
/// # Errors
/// Returns [`MalformedJournal`] describing what was wrong and where.
pub fn decode(line: &str, line_number: usize) -> Result<JournalEvent, MalformedJournal> {
    Parser::new(line, line_number).parse_event()
}

/// Recursive-descent parser for the restricted grammar the journal uses.
///
/// Accepting only what the format actually uses keeps the fuzz surface small
/// and makes every rejection specific. Operates on `char` positions rather than
/// bytes so a multi-byte UTF-8 value cannot be split mid-character.
struct Parser {
    chars: Vec<char>,
    line_number: usize,
    pos: usize,
}

impl Parser {
    fn new(src: &str, line_number: usize) -> Self {
        Self {
            chars: src.chars().collect(),
            line_number,
            pos: 0,
        }
    }

    fn parse_event(&mut self) -> Result<JournalEvent, MalformedJournal> {
        self.skip_whitespace();
        self.expect('{')?;

        let mut fields: Option<BTreeMap<String, String>> = None;
        let mut seq: Option<u64> = None;
        let mut event_type: Option<String> = None;

        self.skip_whitespace();
        if self.peek()? != '}' {
            loop {
                self.skip_whitespace();
                let key = self.parse_string()?;
                self.skip_whitespace();
                self.expect(':')?;
                self.skip_whitespace();
                match key.as_str() {
                    KEY_FIELDS => {
                        self.require_absent(fields.is_some(), KEY_FIELDS)?;
                        fields = Some(self.parse_fields()?);
                    }
                    KEY_SEQ => {
                        self.require_absent(seq.is_some(), KEY_SEQ)?;
                        seq = Some(self.parse_non_negative_u64()?);
                    }
                    KEY_TYPE => {
                        self.require_absent(event_type.is_some(), KEY_TYPE)?;
                        event_type = Some(self.parse_string()?);
                    }
                    other => return Err(self.fail(&format!("unknown key \"{other}\""))),
                }
                self.skip_whitespace();
                if self.peek()? == ',' {
                    self.pos += 1;
                    continue;
                }
                break;
            }
        }

        self.expect('}')?;
        self.skip_whitespace();
        if self.pos != self.chars.len() {
            return Err(self.fail("trailing content after the object"));
        }

        // STUDY: option-vs-nullable
        let fields = fields.ok_or_else(|| self.fail(&format!("missing key \"{KEY_FIELDS}\"")))?;
        let seq = seq.ok_or_else(|| self.fail(&format!("missing key \"{KEY_SEQ}\"")))?;
        let event_type =
            event_type.ok_or_else(|| self.fail(&format!("missing key \"{KEY_TYPE}\"")))?;

        Ok(JournalEvent {
            seq,
            event_type,
            fields,
        })
    }

    fn parse_fields(&mut self) -> Result<BTreeMap<String, String>, MalformedJournal> {
        let mut fields = BTreeMap::new();
        self.expect('{')?;
        self.skip_whitespace();
        if self.peek()? == '}' {
            self.pos += 1;
            return Ok(fields);
        }
        loop {
            self.skip_whitespace();
            let key = self.parse_string()?;
            self.skip_whitespace();
            self.expect(':')?;
            self.skip_whitespace();
            if self.peek()? != '"' {
                return Err(self.fail(&format!("field \"{key}\" must be a string")));
            }
            let value = self.parse_string()?;
            if fields.insert(key.clone(), value).is_some() {
                return Err(self.fail(&format!("duplicate field \"{key}\"")));
            }
            self.skip_whitespace();
            if self.peek()? == ',' {
                self.pos += 1;
                continue;
            }
            break;
        }
        self.expect('}')?;
        Ok(fields)
    }

    fn parse_string(&mut self) -> Result<String, MalformedJournal> {
        self.expect('"')?;
        let mut out = String::new();
        loop {
            let Some(&c) = self.chars.get(self.pos) else {
                return Err(self.fail("unterminated string"));
            };
            self.pos += 1;
            if c == '"' {
                return Ok(out);
            }
            if c != '\\' {
                if (c as u32) < 0x20 {
                    return Err(self.fail("raw control character in string"));
                }
                out.push(c);
                continue;
            }
            let Some(&esc) = self.chars.get(self.pos) else {
                return Err(self.fail("unterminated escape"));
            };
            self.pos += 1;
            match esc {
                '"' => out.push('"'),
                '\\' => out.push('\\'),
                '/' => out.push('/'),
                'b' => out.push('\u{8}'),
                'f' => out.push('\u{c}'),
                'n' => out.push('\n'),
                'r' => out.push('\r'),
                't' => out.push('\t'),
                'u' => out.push(self.parse_unicode_escape()?),
                other => return Err(self.fail(&format!("invalid escape \"\\{other}\""))),
            }
        }
    }

    fn parse_unicode_escape(&mut self) -> Result<char, MalformedJournal> {
        if self.pos + 4 > self.chars.len() {
            return Err(self.fail("truncated unicode escape"));
        }
        let mut value: u32 = 0;
        for i in 0..4 {
            let digit = self.chars[self.pos + i]
                .to_digit(16)
                .ok_or_else(|| self.fail("invalid unicode escape"))?;
            value = value * 16 + digit;
        }
        self.pos += 4;
        char::from_u32(value).ok_or_else(|| self.fail("unicode escape is not a scalar value"))
    }

    fn parse_non_negative_u64(&mut self) -> Result<u64, MalformedJournal> {
        let start = self.pos;
        while self.chars.get(self.pos).is_some_and(char::is_ascii_digit) {
            self.pos += 1;
        }
        if self.pos == start {
            return Err(self.fail(&format!(
                "expected a non-negative integer for \"{KEY_SEQ}\""
            )));
        }
        let digits: String = self.chars[start..self.pos].iter().collect();
        digits
            .parse::<u64>()
            .map_err(|_| self.fail("sequence number out of range"))
    }

    fn require_absent(&self, already_seen: bool, key: &str) -> Result<(), MalformedJournal> {
        if already_seen {
            return Err(self.fail(&format!("duplicate key \"{key}\"")));
        }
        Ok(())
    }

    fn skip_whitespace(&mut self) {
        while matches!(self.chars.get(self.pos), Some(' ' | '\t' | '\r' | '\n')) {
            self.pos += 1;
        }
    }

    fn peek(&self) -> Result<char, MalformedJournal> {
        self.chars
            .get(self.pos)
            .copied()
            .ok_or_else(|| self.fail("unexpected end of line"))
    }

    fn expect(&mut self, expected: char) -> Result<(), MalformedJournal> {
        if self.chars.get(self.pos) != Some(&expected) {
            return Err(self.fail(&format!("expected '{expected}' at offset {}", self.pos)));
        }
        self.pos += 1;
        Ok(())
    }

    fn fail(&self, message: &str) -> MalformedJournal {
        MalformedJournal {
            line: self.line_number,
            message: message.to_owned(),
        }
    }
}

File: cpp/ems-core/src/journal.cpp
#include "ems_core/journal.hpp"

#include <array>
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <iterator>
#include <optional>
#include <sstream>
#include <string>
#include <string_view>
#include <utility>

namespace ems::core {
namespace {

constexpr std::string_view kKeyFields = "fields";
constexpr std::string_view kKeySeq = "seq";
constexpr std::string_view kKeyType = "type";

void append_hex_escape(std::string& out, std::uint32_t code_point) {
    constexpr std::array<char, 16> kHexDigits = {'0', '1', '2', '3', '4', '5', '6', '7',
                                                 '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
    out.append("\\u");
    out.push_back(kHexDigits.at((code_point >> 12U) & 0xFU));
    out.push_back(kHexDigits.at((code_point >> 8U) & 0xFU));
    out.push_back(kHexDigits.at((code_point >> 4U) & 0xFU));
    out.push_back(kHexDigits.at(code_point & 0xFU));
}

void append_string(std::string& out, std::string_view value) {
    out.push_back('"');
    for (const char raw : value) {
        // Bytes >= 0x80 are UTF-8 continuation or lead bytes and are copied
        // through untouched — all three languages agree on UTF-8, and they
        // would not agree on when to escape. The cast keeps the comparisons
        // below from depending on whether char is signed on this platform.
        const auto byte = static_cast<unsigned char>(raw);
        switch (byte) {
            case '"':
                out.append("\\\"");
                break;
            case '\\':
                out.append("\\\\");
                break;
            case '\b':
                out.append("\\b");
                break;
            case '\f':
                out.append("\\f");
                break;
            case '\n':
                out.append("\\n");
                break;
            case '\r':
                out.append("\\r");
                break;
            case '\t':
                out.append("\\t");
                break;
            default:
                if (byte < 0x20U) {
                    // No short escape exists; a raw control byte in the journal
                    // is a byte-level divergence waiting to happen between the
                    // three writers.
                    append_hex_escape(out, byte);
                } else {
                    out.push_back(raw);
                }
                break;
        }
    }
    out.push_back('"');
}

/// Appends `code_point` to `out` as UTF-8.
///
/// Hand-rolled rather than reached for from a library: the journal's only
/// non-ASCII path is \uXXXX unescaping, and pulling in a converter for four
/// branches would be a dependency the byte contract does not need.
void append_utf8(std::string& out, std::uint32_t code_point) {
    if (code_point < 0x80U) {
        out.push_back(static_cast<char>(code_point));
    } else if (code_point < 0x800U) {
        out.push_back(static_cast<char>(0xC0U | (code_point >> 6U)));
        out.push_back(static_cast<char>(0x80U | (code_point & 0x3FU)));
    } else if (code_point < 0x10000U) {
        out.push_back(static_cast<char>(0xE0U | (code_point >> 12U)));
        out.push_back(static_cast<char>(0x80U | ((code_point >> 6U) & 0x3FU)));
        out.push_back(static_cast<char>(0x80U | (code_point & 0x3FU)));
    } else {
        out.push_back(static_cast<char>(0xF0U | (code_point >> 18U)));
        out.push_back(static_cast<char>(0x80U | ((code_point >> 12U) & 0x3FU)));
        out.push_back(static_cast<char>(0x80U | ((code_point >> 6U) & 0x3FU)));
        out.push_back(static_cast<char>(0x80U | (code_point & 0x3FU)));
    }
}

/// Recursive-descent parser for the restricted grammar the journal uses.
///
/// Accepting only what the format actually uses keeps the fuzz surface small
/// and makes every rejection specific. Every read goes through peek()/advance(),
/// which bounds-check — this parser is a fuzz target, so an out-of-range index
/// would be the exact defect the polyglot gate exists to catch.
class Parser {
public:
    Parser(std::string_view src, std::size_t line_number) : src_(src), line_number_(line_number) {}

    Result<JournalEvent, MalformedJournal> parse_event() {
        skip_whitespace();
        if (auto status = expect('{'); !status) {
            return Failure{status.error()};
        }

        std::optional<std::map<std::string, std::string>> fields;
        std::optional<std::uint64_t> seq;
        std::optional<std::string> type;

        skip_whitespace();
        if (pos_ >= src_.size()) {
            return Failure{fail("unexpected end of line")};
        }
        if (src_[pos_] != '}') {
            while (true) {
                skip_whitespace();
                auto key = parse_string();
                if (!key) {
                    return Failure{key.error()};
                }
                skip_whitespace();
                if (auto status = expect(':'); !status) {
                    return Failure{status.error()};
                }
                skip_whitespace();

                if (key.value() == kKeyFields) {
                    if (fields.has_value()) {
                        return Failure{duplicate(kKeyFields)};
                    }
                    auto parsed = parse_fields();
                    if (!parsed) {
                        return Failure{parsed.error()};
                    }
                    fields = parsed.value();
                } else if (key.value() == kKeySeq) {
                    if (seq.has_value()) {
                        return Failure{duplicate(kKeySeq)};
                    }
                    auto parsed = parse_non_negative_u64();
                    if (!parsed) {
                        return Failure{parsed.error()};
                    }
                    seq = parsed.value();
                } else if (key.value() == kKeyType) {
                    if (type.has_value()) {
                        return Failure{duplicate(kKeyType)};
                    }
                    auto parsed = parse_string();
                    if (!parsed) {
                        return Failure{parsed.error()};
                    }
                    type = parsed.value();
                } else {
                    return Failure{fail("unknown key \"" + key.value() + "\"")};
                }

                skip_whitespace();
                if (pos_ < src_.size() && src_[pos_] == ',') {
                    ++pos_;
                    continue;
                }
                break;
            }
        }

        if (auto status = expect('}'); !status) {
            return Failure{status.error()};
        }
        skip_whitespace();
        if (pos_ != src_.size()) {
            return Failure{fail("trailing content after the object")};
        }

        if (!fields.has_value()) {
            return Failure{missing(kKeyFields)};
        }
        if (!seq.has_value()) {
            return Failure{missing(kKeySeq)};
        }
        if (!type.has_value()) {
            return Failure{missing(kKeyType)};
        }

        return JournalEvent{*seq, std::move(*type), std::move(*fields)};
    }

private:
    Result<std::map<std::string, std::string>, MalformedJournal> parse_fields() {
        std::map<std::string, std::string> fields;
        if (auto status = expect('{'); !status) {
            return Failure{status.error()};
        }
        skip_whitespace();
        if (pos_ < src_.size() && src_[pos_] == '}') {
            ++pos_;
            return fields;
        }
        while (true) {
            skip_whitespace();
            auto key = parse_string();
            if (!key) {
                return Failure{key.error()};
            }
            skip_whitespace();
            if (auto status = expect(':'); !status) {
                return Failure{status.error()};
            }
            skip_whitespace();
            if (pos_ >= src_.size() || src_[pos_] != '"') {
                return Failure{fail("field \"" + key.value() + "\" must be a string")};
            }
            auto value = parse_string();
            if (!value) {
                return Failure{value.error()};
            }
            if (!fields.emplace(key.value(), value.value()).second) {
                return Failure{fail("duplicate field \"" + key.value() + "\"")};
            }
            skip_whitespace();
            if (pos_ < src_.size() && src_[pos_] == ',') {
                ++pos_;
                continue;
            }
            break;
        }
        if (auto status = expect('}'); !status) {
            return Failure{status.error()};
        }
        return fields;
    }

    Result<std::string, MalformedJournal> parse_string() {
        if (auto status = expect('"'); !status) {
            return Failure{status.error()};
        }
        std::string out;
        while (true) {
            if (pos_ >= src_.size()) {
                return Failure{fail("unterminated string")};
            }
            const char raw = src_[pos_++];
            const auto byte = static_cast<unsigned char>(raw);
            if (raw == '"') {
                return out;
            }
            if (raw != '\\') {
                if (byte < 0x20U) {
                    return Failure{fail("raw control character in string")};
                }
                out.push_back(raw);
                continue;
            }
            if (pos_ >= src_.size()) {
                return Failure{fail("unterminated escape")};
            }
            const char esc = src_[pos_++];
            switch (esc) {
                case '"':
                    out.push_back('"');
                    break;
                case '\\':
                    out.push_back('\\');
                    break;
                case '/':
                    out.push_back('/');
                    break;
                case 'b':
                    out.push_back('\b');
                    break;
                case 'f':
                    out.push_back('\f');
                    break;
                case 'n':
                    out.push_back('\n');
                    break;
                case 'r':
                    out.push_back('\r');
                    break;
                case 't':
                    out.push_back('\t');
                    break;
                case 'u': {
                    auto code_point = parse_unicode_escape();
                    if (!code_point) {
                        return Failure{code_point.error()};
                    }
                    append_utf8(out, code_point.value());
                    break;
                }
                default:
                    return Failure{fail(std::string("invalid escape \"\\") + esc + "\"")};
            }
        }
    }

    Result<std::uint32_t, MalformedJournal> parse_unicode_escape() {
        if (pos_ + 4U > src_.size()) {
            return Failure{fail("truncated unicode escape")};
        }
        std::uint32_t value = 0;
        for (std::size_t i = 0; i < 4U; ++i) {
            const char c = src_[pos_ + i];
            std::uint32_t digit = 0;
            if (c >= '0' && c <= '9') {
                digit = static_cast<std::uint32_t>(c - '0');
            } else if (c >= 'a' && c <= 'f') {
                digit = static_cast<std::uint32_t>(c - 'a') + 10U;
            } else if (c >= 'A' && c <= 'F') {
                digit = static_cast<std::uint32_t>(c - 'A') + 10U;
            } else {
                return Failure{fail("invalid unicode escape")};
            }
            value = (value * 16U) + digit;
        }
        pos_ += 4U;
        // Surrogate halves are legal JSON escapes but not Unicode scalar
        // values. Rust's char cannot hold one, so accepting it here would put a
        // different byte sequence in the C++ journal than in the Rust journal —
        // exactly the divergence the conformance gate exists to catch.
        if (value >= 0xD800U && value <= 0xDFFFU) {
            return Failure{fail("unicode escape is not a scalar value")};
        }
        return value;
    }

    Result<std::uint64_t, MalformedJournal> parse_non_negative_u64() {
        const std::size_t start = pos_;
        while (pos_ < src_.size() && src_[pos_] >= '0' && src_[pos_] <= '9') {
            ++pos_;
        }
        if (pos_ == start) {
            return Failure{fail("expected a non-negative integer for \"" + std::string(kKeySeq) + "\"")};
        }
        std::uint64_t value = 0;
        for (std::size_t i = start; i < pos_; ++i) {
            const auto digit = static_cast<std::uint64_t>(src_[i] - '0');
            // Overflow must be a rejection, not a wrap: a wrapped sequence
            // number is a journal that replays in a different order.
            if (value > ((UINT64_MAX - digit) / 10U)) {
                return Failure{fail("sequence number out of range")};
            }
            value = (value * 10U) + digit;
        }
        return value;
    }

    Status<MalformedJournal> expect(char expected) {
        if (pos_ >= src_.size() || src_[pos_] != expected) {
            return Failure{fail(std::string("expected '") + expected + "' at offset " +
                                std::to_string(pos_))};
        }
        ++pos_;
        return {};
    }

    void skip_whitespace() {
        while (pos_ < src_.size() &&
               (src_[pos_] == ' ' || src_[pos_] == '\t' || src_[pos_] == '\r' ||
                src_[pos_] == '\n')) {
            ++pos_;
        }
    }

    [[nodiscard]] MalformedJournal fail(std::string message) const {
        return MalformedJournal{line_number_, std::move(message)};
    }

    [[nodiscard]] MalformedJournal duplicate(std::string_view key) const {
        return fail("duplicate key \"" + std::string(key) + "\"");
    }

    [[nodiscard]] MalformedJournal missing(std::string_view key) const {
        return fail("missing key \"" + std::string(key) + "\"");
    }

    std::string_view src_;
    std::size_t line_number_;
    std::size_t pos_{0};
};

}  // namespace

std::string encode(const JournalEvent& event) {
    std::string out;
    out.append("{\"").append(kKeyFields).append("\":{");
    bool first = true;
    for (const auto& [key, value] : event.fields) {
        if (!first) {
            out.push_back(',');
        }
        first = false;
        append_string(out, key);
        out.push_back(':');
        append_string(out, value);
    }
    out.append("},\"").append(kKeySeq).append("\":").append(std::to_string(event.seq));
    out.append(",\"").append(kKeyType).append("\":");
    append_string(out, event.type);
    out.push_back('}');
    return out;
}

Result<JournalEvent, MalformedJournal> decode(std::string_view line, std::size_t line_number) {
    Parser parser(line, line_number);
    return parser.parse_event();
}

Result<std::vector<JournalEvent>, MalformedJournal> read_journal(
    const std::filesystem::path& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        return Failure{MalformedJournal{0, "cannot open " + path.string()}};
    }
    const std::string raw((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());

    std::vector<JournalEvent> events;
    std::size_t line_number = 0;
    std::size_t start = 0;
    while (start <= raw.size()) {
        const std::size_t end = raw.find('\n', start);
        const std::size_t stop = (end == std::string::npos) ? raw.size() : end;
        const std::string_view line(raw.data() + start, stop - start);
        ++line_number;

        const bool blank =
            line.find_first_not_of(" \t\r") == std::string_view::npos;
        if (!blank) {
            auto parsed = decode(line, line_number);
            if (!parsed) {
                return Failure{parsed.error()};
            }
            events.push_back(parsed.value());
        }

        if (end == std::string::npos) {
            break;
        }
        start = end + 1U;
    }
    return events;
}

Status<MalformedJournal> write_journal(const std::filesystem::path& path,
                                       const std::vector<JournalEvent>& events) {
    std::ofstream out(path, std::ios::binary | std::ios::trunc);
    if (!out) {
        return Failure{MalformedJournal{0, "cannot open " + path.string() + " for writing"}};
    }
    for (const auto& event : events) {
        out << encode(event) << '\n';
    }
    out.flush();
    if (!out) {
        return Failure{MalformedJournal{0, "write failed for " + path.string()}};
    }
    return {};
}

}  // namespace ems::core
